Add Nest cam support for the SDM API

allenporter opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add a nest camera platform for the nest integration that works with
the SDM API.  This works similar to sensors in that a pubsub subscriber
listens for updates to traits about the camera.  For the initial step,
this only supports an RTSP stream, and not still images.

In the future, still image support can be added, but only when triggered by
an event on the camera (e.g. motion, person, sound, etc).

This also contains some refactoring for sharing common code between platforms, and in test setup.  The device specific attributes that are common across entities have been pulled info a device_info library.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/15391

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

allenporter committed:
Camera platform for nest integration SDM API

Add a nest camera platform for the nest integration that works with
the SDM API.  This works similar to sensors in that a pubsub subscriber
listens for updates to traits about the camera.  For the initial step,
this only supports an RTSP stream, and not still images.

In the future, still image support can be added, but only when triggered by
an event on the camera (e.g. motion, person, sound, etc).

allenporter committed:
Suppress nest NotImplementedError for camera bytes

Support for still images will come later.

allenporter committed:
Update device manager getter

Update async properties to be a rela method, to match latest
python-google-nest-sdm version

allenporter committed:
Fix bug in nest device info class for unknowns

Add a test that exercises the DeviceInfo class and fixes a bug where
the device name is not known.  It is now inferred from the type, if
possible.

mike9011 commented:
Great work @allenporter ! Unfortunately after trying the new version with camera support, sometimes my Home Assistant becomes unresponsive, while the CPU load is ~30%. Is there anything I can log to see what's happening?

allenporter commented:
Not sure what the etiquette is but happy to discuss here.   When I play a stream, hass is using a tiny amount of cpu, so i 0.1 % cpu so i haven't been able to reproduce your problem.

When I want to turn up loggging selectively I do something like this in my `configuration.yaml`:
```
logger:
  default: info
  logs:
    homeassistant.components.cloud: debug
    homeassistant.components.nest: debug
    homeassistant.components.nest.sensor_sdm: debug
    homeassistant.components.stream: debug
```
Though it assumes you know where you want to probe.  Cranking up the default could work too.

I suspect the 4XX errors are because this doesn't yet support refreshing the stream due to expiration.  Playing around with it myself it seemed to work for a few minutes, then stops until i re-open.  There is a comment about this as a future TODO, though it may be something i should fix in this PR since you can't selectively enable nest cams.

mike9011 commented:
> Not sure what the etiquette is but happy to discuss here. When I play a stream, hass is using a tiny amount of cpu, so i 0.1 % cpu so i haven't been able to reproduce your problem.
> 
> When I want to turn up loggging selectively I do something like this in my `configuration.yaml`:
> 
> ```
> logger:
>   default: info
>   logs:
>     homeassistant.components.cloud: debug
>     homeassistant.components.nest: debug
>     homeassistant.components.nest.sensor_sdm: debug
>     homeassistant.components.stream: debug
> ```
> 
> Though it assumes you know where you want to probe. Cranking up the default could work too.
> 
> I suspect the 4XX errors are because this doesn't yet support refreshing the stream due to expiration. Playing around with it myself it seemed to work for a few minutes, then stops until i re-open. There is a comment about this as a future TODO, though it may be something i should fix in this PR since you can't selectively enable nest cams.
> 
> Overall, i have not yet figured out the entire `camera` lifecycle / state machine and interaction with `stream`. (I may try to dig in and document it if i figure it out)

Looks like the unresponsiveness was a temporary issue. 

Ah ok that makes sense! That's probably why it takes longer to load the stream when I reopen the application. For a second I see the last image of the last session and then the "broken stream icon". 

Thanks again @allenporter really appreciate your work.

allenporter commented:
Comment for reviewers:
- Would you prefer refactor to share code as a "pre-factor" pull request?
- Would like initial feedback to make sure i'm headed in the right direction with this, before adding refreshing the expiration token (any feedback on approach welcome)

allenporter committed:
Fix nest camera return value

This function does not return anything, so just have an empty return.  This
was feedback from PR for the sensor integration, which I didn't copy to the
camera integration.

balloob commented on homeassistant/components/nest/camera_sdm.py:
```diff
@@ -0,0 +1,115 @@
+        self._device_info = DeviceInfo(device)
+
+    @property
+    def should_pool(self) -> bool:
```
```suggestion
    def should_poll(self) -> bool:
```

balloob committed:
Typo

balloob commented on homeassistant/components/nest/camera_sdm.py:
```diff
@@ -0,0 +1,115 @@
+
+    async def async_camera_image(self):
+        """Return bytes of camera image."""
+        # No support for still images yet.  Still images are only available
```
For the future, we can also use FFMPEG to extract a still image from an RTSP stream

https://github.com/home-assistant/core/blob/08342a1e051f588ad7e1249185f4716a6052f52d/homeassistant/components/ffmpeg/camera.py#L53-L63

balloob reviewed (approved):
looks great !
